### PR TITLE
refactor iorg for object-labels

### PR DESCRIPTION
**Pull request by leifdenby**
@martinjanssens I've started this work and as so often is the case it is more complicated than I thought :) The challenge here is around what to do when we're using data with periodic boundary conditions. As you originally wrote the code the `utils.periodic` function created a `(nx*2, ny*2)` copy of the original cloud-mask, identified objects in this mask, then computed the centroids of these objects and moved the centroids for the objects that were "outside" the original mask. This means that refactoring this code isn't as simple as just passing in the labelled-objects 2D array, since this the centroid calculation for example won't work for objects that straddle the boundary.

So this needs a bit more thought...

Completing https://github.com/cloudsci/cloudmetrics/issues/52

### CODE REVIEW

**Comment by martinjanssens** [comment]
I'm probably the slow one here: What's wrong with passing the object labelled array that you make after applying the make_periodic_domain function? That would be what we want to happen for all object-based metrics that have periodic BCs, right (and what is currently happening [here](https://github.com/cloudsci/cloudmetrics/blob/master/cloudmetrics/mask/objects.py)? Could we do that for iorg as well? I think it worked before we started refactoring, as long as I moved the objects whose centroid was outside the original domain after applying`utils.periodic` back into the original domain, cut off the extra space and applied the pairwise distance calculation in a periodic box. I'm probably overlooking something, though!

**Comment by leifdenby** [comment]
Sorry that wasn't clear above. The issue is in assuming all metric functions which operate on labelled objects don't need to know whether the labelled objects are on a periodic domain. If we start off with the following object mask in a periodic domain:

```
.........
.........
M......MM
MM.....MM
.........
..MM.....
```

we'd end up with the following object labels (with object `A` straddling the boundary):

```
.........
.........
A......AA
AA.....AA
.........
..BB.....
```

If we then call the iorg computation on these labelled objects, without also passing through that these labelled objects are defined with a 2D array which has been double in size in x- and y-direction, then iorg calculation will be wrong. You added functionality inside the iorg function to unwrap the object positions before, which is great, but to use it I have to pass an argument in to indicate that the labelled objects array represents objects in a periodic domain. I'm nearly there though I think :)

**Comment by leifdenby** [comment]
AH! I think this is finally ready for review @martinjanssens :partying_face: Hope you're happy with what I've done!

**Review by martinjanssens — COMMENTED** [review]
Wow, wonderful! Thanks a lot, it looks really nice. So for my understanding, just adding an additional check whether each individual metric function operating on `object_labels` takes `periodic_domain` as an input is enough? That's clever :)

**Comment by leifdenby** [comment]
> So for my understanding, just adding an additional check whether each individual metric function operating on `object_labels` takes `periodic_domain` as an input is enough? That's clever :)

Yes :) I finally realised that I could do just that. Eventually I would like to make something better using the `inspect` module to construct mask-based metric functions from the object-based metrics, instead of the hacky `eval`-hack I use now, but this works for now :)



> I think this is ready to be merged in, there's just a few documentation notes for the (internal) function `_compute_inhibition_nearest_neighbour_distribution` that I think are outdated with your new implementation (e.g. we had a `num_placements` argument that allowed you to repeat the placement routine a few times to not make the `iorg` you compute too sensitive to a single random placement procedure, whose statistics might not be entirely converged; it is still being explained in the docstring). Should we just clean that up now, so we don't forget about it later?

Ah, I missed this. I'll fix the docstring and merge :) thanks!

File: cloudmetrics/objects/metrics/iorg.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+    # Gather positions in array
+    pos_rand = np.zeros((len(placed_circles), 2))
+    for i, placed_circle in enumerate(placed_circles):
+        pos_rand[i, 0] = placed_circle.x
+        pos_rand[i, 1] = placed_circle.y
+
+    # If field has open bcs, do not compute nn distances using
+    # periodic bcs
+    nnd_rand = find_nearest_neighbors(pos_rand, nn_window)
```
Comment (leifdenby) [inline cloudmetrics/objects/metrics/iorg.py:176]: Actually @martinjanssens, I do have a question here: Do we need to restrict the the maximum distance for the nearest-neighbour search periodic domains? The wrapping you wrote that places all centroids inside the domain should ensure that it is impossible to have any nearest-neighbour pair further apart than the domain size.

Another question I realised while typing this: If I have a set of centroids (two objects) placed like this

```
[      ]
................
................
......x..x......
................
................
................
................
................
................
................
................
................
................
................
................
................
```

After wrapping so the centroids are inside the domain that will result in:

```
[      ]
........
........
.x....x.
........
........
........
........
........
```

Meaning that the wrapping causes these points to move further apart (!) Have a got something wrong here or could this happen?

Comment (martinjanssens) [reply cloudmetrics/objects/metrics/iorg.py:176]: Really great questions, I remember spending quite some time thinking about these things! The way I understand the `KDTree` method operating inside `find_nearest_neighbours` to work, it should handle all this when you give it the `boxsize` argument, i.e. it realises that it should always use the smallest distance between the points to compute the distance to the nearest neighbour and uses this when it's specified (see https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.KDTree.html). So I think that together with all the points being inside the domain and the domainsize being restricted to half the size of the input mask, everything should work consistently. I remember doing some tests that showed as much too, but admit I can't show anything right away. I'm curious what you think!

Comment (leifdenby) [reply cloudmetrics/objects/metrics/iorg.py:176]: Thanks for this @martinjanssens! I've created a test to check that the nearest-neighbor distance calculation indeed works for periodic domains (https://github.com/cloudsci/cloudmetrics/pull/61). If you could give that a quick glance that would be great. I'd like to get that merged in, and then I'll merge this pull-request in too, and we'll be ready for our first release!

**Pull request merged**